### PR TITLE
[Feature] Support Label Assignment Distillation (LAD)

### PR DESCRIPTION
**Pull request by thuyngch**
## Motivation

Add a new label assignment method [LAD](https://github.com/cybercore-co-ltd/CoLAD) into mmdetection.

## Modification

- LAD detector inherits from `KnowledgeDistillationSingleStageDetector`.
- LAD head inherits from `PAAHead`.

## Checklist

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by hhaAndroid** [comment]
> Hi alls, I use the code to train two models in the `configs/lad/README.md` file and can reproduce results in the paper. I would like to share checkpoints and training logs here: https://www.dropbox.com/sh/kdqp1zyt5vtjqzz/AABJ1j_xUd--p_rwzJ3ZZ86ca?dl=0 Thank you!

Thank you for your contribution, we will organize a review immediately.

File: mmdet/models/detectors/lad.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+                 teacher_bbox_head,
+                 teacher_ckpt,
+                 eval_teacher=True,
+                 train_cfg=None,
+                 test_cfg=None,
+                 pretrained=None):
+        super(KnowledgeDistillationSingleStageDetector,
+              self).__init__(backbone, neck, bbox_head, train_cfg, test_cfg,
+                             pretrained)
+        self.eval_teacher = eval_teacher
```
Comment (hhaAndroid) [inline mmdet/models/detectors/lad.py:28]: No need, pass directly to the parent class.

File: mmdet/models/detectors/lad.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            gt_bboxes (list[Tensor]): Each item are the truth boxes for each
+                image in [tl_x, tl_y, br_x, br_y] format.
+            gt_labels (list[Tensor]): Class indices corresponding to each box
+            gt_bboxes_ignore (None | list[Tensor]): Specify which bounding
+                boxes can be ignored when computing the loss.
+
+        Returns:
+            dict[str, Tensor]: A dictionary of loss components.
+        """
+        with torch.no_grad():
```
Comment (hhaAndroid) [inline mmdet/models/detectors/lad.py:77]: If this means it must be `eval_teacher=True`? If so, you need to add some assert.

File: mmdet/models/dense_heads/lad_head.py [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+from .paa_head import PAAHead, levels_to_images
+
+
+@HEADS.register_module()
+class LADHead(PAAHead):
+    """Label Assignment Head from the paper: `Improving Object Detection by
+    Label Assignment Distillation <https://arxiv.org/pdf/2108.10520.pdf>`_"""
+
+    @force_fp32(apply_to=('cls_scores', 'bbox_preds', 'iou_preds'))
+    def get_la(self,
```
Comment (hhaAndroid) [inline mmdet/models/dense_heads/lad_head.py:15]: You can use the full name directly, such as `label_assignment_distillation()`.

File: mmdet/models/dense_heads/lad_head.py [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+        outs = self(x)
+        if gt_labels is None:
+            loss_inputs = outs + (gt_bboxes, img_metas)
+        else:
+            loss_inputs = outs + (gt_bboxes, gt_labels, img_metas)
+        losses = self.loss(
+            *loss_inputs,
+            gt_bboxes_ignore=gt_bboxes_ignore,
+            la_results=la_results)
+        if proposal_cfg is None:
```
Comment (hhaAndroid) [inline mmdet/models/dense_heads/lad_head.py:160]: When will proposal_cfg be used?

File: mmdet/models/dense_heads/lad_head.py (hunk 3 above)
Comment (RangiLyu) [inline mmdet/models/dense_heads/lad_head.py:15]:
```suggestion
    def get_label_assign(self,
```
The full name would be easier to understand.

File: mmdet/models/dense_heads/lad_head.py [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+               bbox_preds,
+               iou_preds,
+               gt_bboxes,
+               gt_labels,
+               img_metas,
+               gt_bboxes_ignore=None):
+        """Get label assignment (from teacher).
+
+        Args:
+            cls_scores (list[Tensor]): Box scores for each scale level
```
Comment (RangiLyu) [inline mmdet/models/dense_heads/lad_head.py:26]:
```suggestion
            cls_scores (list[Tensor]): Box scores for each scale level.
```

File: mmdet/models/dense_heads/lad_head.py (hunk 3 above)
Comment (RangiLyu) [reply mmdet/models/dense_heads/lad_head.py:15]: And the return value name is `label_reassign`, should the function be named as `get_label_reassign`?

File: mmdet/models/detectors/lad.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+            x_teacher = self.extract_teacher_feat(img)
+            outs_teacher = self.teacher_model.bbox_head(x_teacher)
+            la_results = self.teacher_model.bbox_head.get_la(
```
Comment (hhaAndroid) [inline mmdet/models/detectors/lad.py:80]: label_reassign_targets

File: mmdet/models/dense_heads/lad_head.py [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+                      gt_labels=None,
+                      gt_bboxes_ignore=None,
+                      proposal_cfg=None,
+                      **kwargs):
+        """Forward train with the available label assignment (student receives
+        from teacher).
+
+        Args:
+            x (list[Tensor]): Features from FPN.
+            la_results (tuple): As the outputs defined in the function
```
Comment (RangiLyu) [inline mmdet/models/dense_heads/lad_head.py:133]: Switch to `label_assign_results` or `label_reassign_results`.

File: mmdet/models/dense_heads/lad_head.py [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+    @force_fp32(apply_to=('cls_scores', 'bbox_preds', 'iou_preds'))
+    def loss(self,
+             cls_scores,
+             bbox_preds,
+             iou_preds,
+             gt_bboxes,
+             gt_labels,
+             img_metas,
+             gt_bboxes_ignore=None,
+             la_results=None):
```
Comment (RangiLyu) [inline mmdet/models/dense_heads/lad_head.py:175]: Same as above

File: mmdet/models/detectors/lad.py (hunk 6 above)
Comment (RangiLyu) [inline mmdet/models/detectors/lad.py:80]: Same as above

**Comment by hhaAndroid** [comment]
@thuyngch I think the code quality is very good. You may need to add unit tests and try not to use abbreviations for variable names.

File: mmdet/models/detectors/lad.py (hunk 1 above)
Comment (thuyngch) [reply mmdet/models/detectors/lad.py:28]: The parent class builds the teacher network via a config file:
```python
        self.teacher_model = build_detector(teacher_config['model'])
```
But my class builds the teacher network via components:
```python
        self.teacher_model = nn.Module()
        self.teacher_model.backbone = build_backbone(teacher_backbone)
        if teacher_neck is not None:
            self.teacher_model.neck = build_neck(teacher_neck)
        teacher_bbox_head.update(train_cfg=train_cfg)
        teacher_bbox_head.update(test_cfg=test_cfg)
        self.teacher_model.bbox_head = build_head(teacher_bbox_head)
```
That's why I re-write the `__init__` function. Do you have any better suggestions?

File: mmdet/models/detectors/lad.py (hunk 2 above)
Comment (thuyngch) [reply mmdet/models/detectors/lad.py:77]: I just naively follow the parent-class code:
```python
        x = self.extract_feat(img)
        with torch.no_grad():
            teacher_x = self.teacher_model.extract_feat(img)
            out_teacher = self.teacher_model.bbox_head(teacher_x)
        losses = self.bbox_head.forward_train(x, out_teacher, img_metas,
                                              gt_bboxes, gt_labels,
                                              gt_bboxes_ignore)
        return losses
```
I think that `eval` operation only freezes BN and Dropout layers, but it still allows backward gradients. So the operation `with torch.no_grad()` is needed to prevent backward gradients to the teacher network.

File: mmdet/models/dense_heads/lad_head.py (hunk 3 above)
Comment (thuyngch) [reply mmdet/models/dense_heads/lad_head.py:15]: Thanks, I will change the name to `get_label_assignment`.

File: mmdet/models/dense_heads/lad_head.py (hunk 4 above)
Comment (thuyngch) [reply mmdet/models/dense_heads/lad_head.py:160]: Thanks, I just naively follow the code in `BaseDenseHead`.
But I think `proposal_cfg` is used for `RPNHead`.
So I will remove this option in `LADHead`.

File: mmdet/models/detectors/lad.py (hunk 6 above)
Comment (thuyngch) [reply mmdet/models/detectors/lad.py:80]: Thanks, I will change the name to `label_assignment_results`.

File: mmdet/models/dense_heads/lad_head.py (hunk 3 above)
Comment (thuyngch) [reply mmdet/models/dense_heads/lad_head.py:15]: Thanks, I will change the function name to `get_label_assignment` and the return variable name to `label_assignment_results`

File: mmdet/models/dense_heads/lad_head.py (hunk 5 above)
Comment (thuyngch) [reply mmdet/models/dense_heads/lad_head.py:26]: Thank you, I will add the dot.

File: mmdet/models/dense_heads/lad_head.py (hunk 7 above)
Comment (thuyngch) [reply mmdet/models/dense_heads/lad_head.py:133]: Thanks, I will change it to `label_assignment_results`.

File: mmdet/models/dense_heads/lad_head.py (hunk 8 above)
Comment (thuyngch) [reply mmdet/models/dense_heads/lad_head.py:175]: Thanks, I will change it to label_assignment_results.

File: mmdet/models/detectors/lad.py (hunk 6 above)
Comment (thuyngch) [reply mmdet/models/detectors/lad.py:80]: Thanks, I will change it to `label_assignment_results`.

**Comment by thuyngch** [comment]
Hi @hhaAndroid and @RangiLyu ,
Thanks for feedbacks from you guys.
I push a commit to avoid abbreviations in variable names.
Now I am writing unit tests.

**Comment by thuyngch** [comment]
Hi @hhaAndroid and @RangiLyu ,
I added the unit test for LADHead.
Is it enough or needed to add more?

**Review by hhaAndroid — COMMENTED** [review]
I think it's okay

**Comment by RangiLyu** [comment]
Hi, we have fixed the compatibility with MMCV in v2.18.1. Now the CI can work normally. Please pull the latest master and re-run the unit tests.

**Review by RangiLyu — APPROVED** [review]
LGTM

File: tests/test_models/test_dense_heads/test_lad_head.py [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+            max_per_img=100))
+    rescale = False
+    self.get_bboxes(
+        cls_scores, bbox_preds, iou_preds, img_metas, cfg, rescale=rescale)
+
+
+# ------------------------------------------------------------------------------
+#  Main execution
+# ------------------------------------------------------------------------------
+if __name__ == '__main__':
```
Comment (hhaAndroid) [inline tests/test_models/test_dense_heads/test_lad_head.py:155]: delete

Comment (thuyngch) [reply tests/test_models/test_dense_heads/test_lad_head.py]: I deleted it

**Review by ZwwWayne — APPROVED** [review]

**Review by hhaAndroid — APPROVED** [review]

**Comment by hhaAndroid** [comment]
@thuyngch I ran `configs/lad/lad_r50_paa_r101_fpn_coco_1x.py`, and found that the performance is 38.4, there should be problems with the code, can you check it?

**Comment by thuyngch** [comment]
I will check it soon.

**Comment by thuyngch** [comment]
> Hi alls,
> I use the code to train two models in the `configs/lad/README.md` file and can reproduce results in the paper.
> I would like to share checkpoints and training logs here: https://www.dropbox.com/sh/kdqp1zyt5vtjqzz/AABJ1j_xUd--p_rwzJ3ZZ86ca?dl=0
> Thank you!

I downloaded checkpoint from this site, then run the evaluation with the following command, and I can reproduce the result:
```bash
#!/usr/bin/env bash
set -e
export GPUS=2
export CUDA_VISIBLE_DEVICES=0,1

CFG="configs/lad/lad_r50_paa_r101_fpn_coco_1x.py"
CKPT="work_dirs/epoch_12-c2880474.pth"
tools/dist_test.sh $CFG $CKPT $GPUS --fuse-conv-bn --eval bbox

CFG="configs/lad/lad_r101_paa_r50_fpn_coco_1x.py"
CKPT="work_dirs/epoch_12-1d8d48d4.pth"
tools/dist_test.sh $CFG $CKPT $GPUS --fuse-conv-bn --eval bbox
```

The evaluation output for `configs/lad/lad_r50_paa_r101_fpn_coco_1x.py`:
```
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.416
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=1000 ] = 0.598
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=1000 ] = 0.456
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=1000 ] = 0.239
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=1000 ] = 0.455
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=1000 ] = 0.550
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.597
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=300 ] = 0.597
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=1000 ] = 0.597
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=1000 ] = 0.398
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=1000 ] = 0.647
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=1000 ] = 0.760
```

The evaluation output for `configs/lad/lad_r101_paa_r50_fpn_coco_1x.py`:
```
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.432
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=1000 ] = 0.615
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=1000 ] = 0.472
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=1000 ] = 0.254
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=1000 ] = 0.477
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=1000 ] = 0.568
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.610
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=300 ] = 0.610
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=1000 ] = 0.610
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=1000 ] = 0.406
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=1000 ] = 0.662
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=1000 ] = 0.772
```

**Comment by thuyngch** [comment]
Ah I just found that in my config file, I use 8 imgs/gpu and I train the model with 2 GPUs.
https://github.com/open-mmlab/mmdetection/blob/dev-v2.19.0/configs/lad/lad_r50_paa_r101_fpn_coco_1x.py#L118

I guest that you use 8 GPUs to train the model, is it right?

**Comment by hhaAndroid** [comment]
> Ah I just found that in my config file, I use 8 imgs/gpu and I train the model with 2 GPUs. https://github.com/open-mmlab/mmdetection/blob/dev-v2.19.0/configs/lad/lad_r50_paa_r101_fpn_coco_1x.py#L118
> 
> I guess that you use 8 GPUs to train the model, is it right?

oK
